src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -0,0 +1,133 @@
+
+    internal interface IStateTable
+    {
+        IStateTable Compact();
+    }
+
+    /// <summary>
+    /// A data structure that tracks the inputs and output of an execution node
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    internal sealed class NodeStateTable<T> : IStateTable
+    {
+        internal static NodeStateTable<T> Empty { get; } = new NodeStateTable<T>(ImmutableArray<ImmutableArray<(T, EntryState)>>.Empty, isCompacted: true, exception: null);
+
+        //PROTOTYPE: there is no need to store the state per item
+        //           we can instead store one state per input, with
+        //           an optional set for modified states.
+        readonly ImmutableArray<ImmutableArray<(T item, EntryState state)>> _states;
+
+        private readonly bool _isCompacted;
+
+        readonly Exception? _exception;
+
+        private NodeStateTable(ImmutableArray<ImmutableArray<(T, EntryState)>> states, bool isCompacted, Exception? exception)
+        {
+            _states = states;
+            _isCompacted = isCompacted;
+            _exception = exception;
+        }
+
+        public bool IsFaulted { get => _exception is not null; }
+
+        public IEnumerator<(T item, EntryState state)> GetEnumerator()
+        {
+            return _states.SelectMany(s => s).GetEnumerator();
+        }
+
+        public IStateTable Compact()
+        {
+            if (_isCompacted)
+                return this;
+
+            var compacted = ArrayBuilder<ImmutableArray<(T, EntryState)>>.GetInstance();
+            foreach (var entry in _states)
+            {
+                // we have to keep empty entries
+                // we only remove all entries at once, so only need to check the first item
+                if (entry.Length == 0 || entry[0].state != EntryState.Removed)
+                {
+                    compacted.Add(entry.SelectAsArray(e => (e.item, EntryState.Cached)));
+                }
+            }
+            return new NodeStateTable<T>(compacted.ToImmutableAndFree(), isCompacted: true, _exception);
+        }
+
+        public static NodeStateTable<T> FromFaultedTable<U>(NodeStateTable<U> table)
+        {
+            Debug.Assert(table.IsFaulted);
+            return new NodeStateTable<T>(Empty._states, isCompacted: true, table._exception);
+        }
+
+        public class Builder
+        {
+            private readonly ArrayBuilder<ImmutableArray<(T, EntryState)>> _states = ArrayBuilder<ImmutableArray<(T, EntryState)>>.GetInstance();
+
+            private Exception? _exception = null;
+
+            public void AddEntries(ImmutableArray<T> values, EntryState state)
+            {
+                _states.Add(values.SelectAsArray(v => (v, state)));
+            }
+
+            public void AddEntriesFromPreviousTable(NodeStateTable<T> previousTable, EntryState newState)
+            {
+                Debug.Assert(previousTable._states.Length > _states.Count);
```

Comment:
Why is it okay assume that previous tables will be larger than the current one?

Reply:
It turns out that it isn't. This was made under the assumption that we always populate the table from the upstream, so there is always going to be data in the previous table. However this assumption breaks down if a new node is added to the graph later on (i.e. the IDE adds a new generator), so we can longer take it as an invariant.

Reply:
Added a proto comment to track this. It's changed slightly in a future PR anyway, but there are some good cases to consider + test, so want to make sure we don't lose track of them.